File: backend/src/main/java/io/github/daegwonkim/backend/common/exception/ErrorResponse.java
package io.github.daegwonkim.backend.common.exception;

import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;

@Builder
public record ErrorResponse(
        int code,
        String message,
        LocalDateTime timestamp,
        String path,

        // 유효성 검증 실패 시 필드별 에러
        List<FieldError> errors
) {
    public record FieldError(
            String field,
            String value,
            String reason
    ) {}

    // 기본 에러 응답
    public static ErrorResponse of(ErrorCode errorCode, String path) {
        return new ErrorResponse(
                errorCode.getCode(),
                errorCode.getMessage(),
                LocalDateTime.now(),
                path,
                null
        );
    }

    // Validation 에러 응답
    public static ErrorResponse of(ErrorCode errorCode, String path, List<FieldError> errors) {
        return new ErrorResponse(
                errorCode.getCode(),
                errorCode.getMessage(),
                LocalDateTime.now(),
                path,
                errors
        );
    }
}

File: backend/src/main/kotlin/io/github/daegwonkim/backend/common/exception/data/ErrorResponse.kt
package io.github.daegwonkim.backend.common.exception.data

import java.time.LocalDateTime


data class ErrorResponse(
    val code: String,
    val message: String,
    val timestamp: LocalDateTime,
    val path: String,
    val errors: List<FieldError>?
) {
    data class FieldError(
        val field: String,
        val value: String,
        val reason: String?
    )

    companion object {
        fun of(errorCode: ErrorCode, path: String): ErrorResponse = ErrorResponse(
            code = errorCode.code,
            message = errorCode.message,
            timestamp = LocalDateTime.now(),
            path = path,
            errors = null
        )

        fun of(errorCode: ErrorCode, path: String, errors: List<FieldError>): ErrorResponse = ErrorResponse(
            code = errorCode.code,
            message = errorCode.message,
            timestamp = LocalDateTime.now(),
            path = path,
            errors = errors
        )
    }
}
